KAFKA-14327: Unify KRaft snapshot generation between broker and controller

cmccabe opened the pull request:
Unify KRaft metadata snapshot generation between the broker and the controller, so that it all flows through the MetadataImage#write code path.

Previously, the broker generated MetadataImage objects inside BrokerMetadataListener, and published them to a single publisher, named BrokerMetadataPublisher. Now, with this PR, MetadataImage objects are loaded by the MetadataLoader, a generic metadata class shared between controller and broker. The image objects are now published to a list of publishers, rather than just to a single publisher.  Also, publishers can also be dynamically added and removed. On the broker side, we use the dynamic addition and removal to only add BrokerMetadataPublisher when our metadata has "caught up" to the controller's.

With this PR, snapshot generation for both the controller and the broker is handled by a metadata publisher named SnapshotGenerator. This replaces BrokerMetadataSnapshotter and the controller's SnapshotGenerator. This means that the code for generating snapshots from the controller data structures can be deleted and we can maintain only one code path here. The new SnapshotGenerator implements time-based snapshotting as specified by KIP-876.

Using the aforementioned image publisher framework, this PR adds support for dynamic configuration
to the controller. Currently, only socket server, authorizer, and metrics reporter reconfiguration
is supported, but more will be added in the future.

This PR also changes how we handle changing the metadata version in MetadataDelta and MetadataImage. Previously, we called handleMetadataVersionChange on all delta objects whenever we encountered a metadata change in the log. This approach was difficult to maintain, however, since changing metadata versions can add or remove fields from both deltas and images. In practice, we did not correctly implement version changes, which is why we had to disable KRaft downgrades in the 3.3 release. The new approach here is to perform an image save and load cycle. That way we only have to maintain one code path for upgrades and downgrades, in MetadataImage#write. Because doing a save and load cycle is somewhat out of scope for MetadataDelta itself, it simply throws a MetadataVersionChangedException and lets the MetadataLoader handle the change.

On a related note, this PR removes offset and epoch tracking from MetadataDelta. This tracking is done by external classes (such as MetadataLoader) so it doesn't need to be done in MetadataDelta. It also introduces a builder for MetadataDelta so that we can add new constructor parameters more easily in the future. (The only new parameter here is MetadataVersion.) Finally, there is a new class, MetadataProvenance, which encompasses offset, epoch, and timestamp, which identifies the source (provenance) of all MetadataImages.

cmccabe committed:
KAFKA-14327: Unify KRaft snapshot generation between broker and controller

Unify KRaft metadata snapshot generation between the broker and the controller, so that it all
flows through the MetadataImage#write code path.

Previously, the broker generated MetadataImage objects inside BrokerMetadataListener, and published
them to a single publisher, named BrokerMetadataPublisher. Now, with this PR, MetadataImage objects
are loaded by the MetadataLoader, a generic metadata class shared between controller and broker.
The image objects are now published to a list of publishers, rather than just to a single
publisher.  Also, publishers can also be dynamically added and removed. On the broker side, we use
the dynamic addition and removal to only add BrokerMetadataPublisher when our metadata has "caught
up" to the controller's.

With this PR, snapshot generation for both the controller and the broker is handled by a metadata
publisher named SnapshotGenerator. This replaces BrokerMetadataSnapshotter and the controller's
SnapshotGenerator. This means that the code for generating snapshots from the controller data
structures can be deleted and we can maintain only one code path here. The new SnapshotGenerator
implements time-based snapshotting as specified by KIP-876.

Using the aforementioned image publisher framework, this PR adds support for dynamic configuration
to the controller. Currently, only socket server, authorizer, and metrics reporter reconfiguration
is supported, but more will be added in the future.

This PR also changes how we handle changing the metadata version in MetadataDelta and
MetadataImage. Previously, we called handleMetadataVersionChange on all delta objects whenever we
encountered a metadata change in the log. This approach was difficult to maintain, however, since
changing metadata versions can add or remove fields from both deltas and images. In practice, we
did not correctly implement version changes, which is why we had to disable KRaft downgrades in the
3.3 release. The new approach here is to perform an image save and load cycle. That way we only
have to maintain one code path for upgrades and downgrades, in MetadataImage#write. Because doing
a save and load cycle is somewhat out of scope for MetadataDelta itself, it simply throws a
MetadataVersionChangedException and lets the MetadataLoader handle the change.

On a related note, this PR removes offset and epoch tracking from MetadataDelta. This tracking is
done by external classes (such as MetadataLoader) so it doesn't need to be done in MetadataDelta.
It also introduces a builder for MetadataDelta so that we can add new constructor parameters more
easily in the future. (The only new parameter here is MetadataVersion.) There is a new class,
MetadataProvenance, which encompasses offset, epoch, and timestamp, which identifies the source
(provenance) of all MetadataImages. This is tracked externally from MetadataDelta, however.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -467,36 +467,41 @@ public void testIterator(MetadataVersion metadataVersion) throws Exception {
                         setHost("example.com")).iterator())).
-                setFenced(true), expectedVersion))),
-                clusterControl.iterator(Long.MAX_VALUE));
+                    setFenced(true), expectedVersion),
```
nit: looks misaligned

hachikuji commented on server-common/src/test/java/org/apache/kafka/queue/KafkaEventQueueTest.java:
```diff
@@ -240,4 +242,29 @@ public void handleException(Throwable e) {
+        assertTrue(queue.isEmpty());
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        queue.append(() -> future.get());
+        assertFalse(queue.isEmpty());
```
This assertion is flaky locally. Can we use `waitForCondition` for all cases where we expect the queue to become empty?

cmccabe replied:
[ I didn't forget this comment -- let me look into it tomorrow morning ]

cmccabe replied:
This had a bug, which is that it wasn't taking into account the currently running event in its calculation of "empty". I fixed this bug and the test should be working now. I don't think we need the waits (except in the one position)

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+        MetadataProvenance provenance =
+                new MetadataProvenance(lastOffset, lastEpoch, lastContainedLogTimeMs);
+        long elapsedNs = time.nanoseconds() - startNs;
+        // TODO: this metric should be renamed something like "delta processing time"
```
nit: can we move this comment to a jira?

cmccabe replied:
OK. I filed [KAFKA-14348](https://issues.apache.org/jira/browse/KAFKA-14348) for this.

I will also remove this comment ... it's probably a distraction having it here

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -726,6 +728,7 @@ object KafkaConfig {
   val MetadataLogDirDoc = "This configuration determines where we put the metadata log for clusters in KRaft mode. " +
     "If it is not set, the metadata log is placed in the first log directory from log.dirs."
   val MetadataSnapshotMaxNewRecordBytesDoc = "This is the maximum number of bytes in the log between the latest snapshot and the high-watermark needed before generating a new snapshot."
+  val MetadataSnapshotMaxIntervalMsDoc = "This is the maximum number of milliseconds that can elapse between the previous metadata snapshot and the next one."
```
Perhaps add something like this?

> In other words, metadata snapshots will be generated at least as frequently as specified by this configuration. Note that if `metadata.log.max.record.bytes.between.snapshots` bytes of new data is reached before the full specified interval completes, then the snapshot will still be generated.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -269,6 +270,26 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
+      addReconfigurable(new DynamicMetricsReporters(
+        kafkaConfig.brokerId, controller.config, controller.metrics, controller.clusterId))
+    }
+    // TODO: add dynamic thread pool resizing here
```
nit: create jiras. We need one for the quota callback as well.

cmccabe replied:
Good point. I created KAFKA-14349, KAFKA-14350, and KAFKA-14351. I will also add comments about them here.

KAFKA-14351 is for controller mutation quotas as a whole (I thought we had that already but I didn't see it, so I created it and linked the appropriate places.)

hachikuji commented on core/src/main/scala/kafka/server/ControllerServer.scala:
```diff
@@ -258,6 +264,14 @@ class ControllerServer(
+      config,
+      dynamicConfigHandlers,
+      fatalFaultHandler)
+    metadataLoader.installPublishers(List().asJava)
```
Don't we need the publisher in here? Do we have any integration tests which verify this all works?

cmccabe replied:
Yes, good catch. I am adding more tests for this (maybe not tonight but tomorrow morning for sure)

hachikuji commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -191,21 +191,36 @@ public KafkaClusterTestKit build() throws Exception {
+                    } catch (Throwable e) {
+                        log.error("Error creating controller {}", node.id(), e);
+                        metadataLoader.close();
+                        controller.shutdown();
```
There should probably be a null check. Same for `broker` below.

cmccabe replied:
good catch

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+    /**
+     * The current leader and epoch.
+     */
+    private LeaderAndEpoch currentLeaderAndEpoch = LeaderAndEpoch.UNKNOWN;
```
This field is not accessed. Do we need it?

cmccabe replied:
it would be nice to keep to minimize divergence (although I don't feel strongly)

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+                    faultHandler.handleFault("Error loading metadata log record from offset " +
+                            batch.baseOffset() + indexWithinBatch, e);
+                }
+                lastEpoch = batch.epoch();
```
nit: pull these two lines outside the loop since they do not depend on the record

cmccabe replied:
This was originally so that we could get the pre-change "provenance" (offset, epoch, timestamp) for writing the snapshot-on-mv-change, if that was needed.

However, I think I found a better way to do this that doesn't require updating those variables on every loop iteration.

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/LogDeltaManifest.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * The number of batches that were loaded.
+     */
+    private final int numBatches;
```
Seems we didn't need this and `elapsedNs` below?

cmccabe replied:
Sorry, this information was supposed to be logged at DEBUG level like in the old loader, but I forgot. I'll add it.

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+            Batch<ApiMessageAndVersion> batch = reader.next();
+            for (ApiMessageAndVersion record : batch.records()) {
+                try {
+                    delta.replay(record.message());
```
This may still throw `MetadataVersionChangeException`, but we are not catching it. Am I missing something?

cmccabe replied:
No, this was a bug. Good catch (btw, I fixed it today while testing some unrelated things)

The scenario where you might hit this is if you had an existing image in memory, and you loaded a snapshot for whatever reason, and the snapshot changed the MV.

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+    public CompletableFuture<Void> installPublishers(List<? extends MetadataPublisher> newPublishers) {
+        if (newPublishers.isEmpty()) return CompletableFuture.completedFuture(null);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        eventQueue.beginShutdown("installPublishers", () -> {
```
Should be `append`, right? Same for `removeAndClosePublisher`. Do we have tests that are failing?

cmccabe replied:
I have now added a MetadataLoaderTest which helped catch and fix a lot of these things.

I forgot to add it the first time around (I had one in some of the draft PRs, but while preparing the final PR I forgot to port it over) :X

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java:
```diff
@@ -0,0 +1,118 @@
+            throw e;
+        } finally {
+            try {
+                snapshotWriter.get().close();
```
This would be clearer without using the try-with-resources, which would also end up closing the writer. I do not know which logic will execute first. Perhaps it's best not to depend on the order.

cmccabe replied:
Yeah, that's a fair point. I'm sure that this ordering is defined by Java, but it's a bit esoteric.

I was assuming that the close order didn't matter, but perhaps it does (or will eventually?) Best just to be straightforward about it. Since they're both AutoCloseable, I will also use `Utils.closeQuietly` as well.

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Override
+    public void emit(MetadataImage image) {
```
nit: maybe we could call this `maybeEmit` or return a boolean or something. Just to make it clear that a snapshot may or may not be emitted by the call.

cmccabe replied:
Fair enough. I will also add some JavaDoc to `SnapshotGenerator.Emitter`

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java:
```diff
@@ -0,0 +1,262 @@
+        }
+
+        public SnapshotGenerator build() {
+            if (emitter == null) throw new RuntimeException("You must set a snapshot emitter.");
```
nit: if this is required, perhaps we should pass it in the constructor

cmccabe replied:
Typically I don't add mandatory params to builder constructors because part of the"value add" builders provide is a visually nicer way of identifying (essentially) constructor parameters. Basically, since we don't have named parameters in Java, we can sort of fake it this way. A mandatory parameter may eventually also become optional, over time.

We wouldn't want to have builders for every object since there is some overhead, both in runtime and visual terms. But for big fat objects (which you typically don't have a lot of anyway), it really does make life a lot nicer.

[ Side note: I really regret not creating builders on day 1 for BrokerServer and ControllerServer since changing the constructor parameters is now an exercise in suffering, even with Scala's named parameters. But I never seem to find a good time to do that refactor. ]

Anyway, in this case, I'll add it, since a SnapshotGenerator without an emitter would be silly.

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java:
```diff
@@ -0,0 +1,262 @@
+                scheduleEmit("we have replayed at least " + minBytesSinceLastSnapshot +
+                    " bytes", newImage);
+            }
+        } else if (minTimeSinceLastSnapshotNs != 0 &&
```
I thought the intent was to disable time-based snapshots with -1.

cmccabe replied:
Hmm [KIP-876](https://cwiki.apache.org/confluence/display/KAFKA/KIP-876%3A+Time+based+cluster+metadata+snapshots) currently says 0 means disabled ...

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java:
```diff
@@ -0,0 +1,262 @@
+        LogContext logContext = new LogContext("[SnapshotGenerator " + nodeId + "] ");
+        this.log = logContext.logger(SnapshotGenerator.class);
+        this.disabledReason = null;
+        this.eventQueue = new KafkaEventQueue(time, logContext, "");
```
Perhaps we should add a thread name prefix?

cmccabe replied:
fair enough. we have all the information here, so there is no reason why not (and some good reasons why to!)

hachikuji commented on core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala:
```diff
@@ -1,413 +0,0 @@
-  }
-
-  @Test
-  def testNotSnapshotAfterMetadataVersionChangeBeforePublishing(): Unit = {
```
Are all of these tests recreated somewhere?

cmccabe replied:
in the latest PR, `MetadataLoaderTest.java` replaces `BrokerMetadataListenerTest.scala`

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/SnapshotManifest.java:
```diff
@@ -0,0 +1,77 @@
+        return provenance;
+    }
+
+    public long elapsedNs() {
```
I don't think this is used. Can we remove it?

cmccabe replied:
similarly to the log delta one, we need it for DEBUG log4j logging (which was missing, but I will add)

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+        int preChangeEpoch,
+        long lastContainedLogTimeMs
+    ) {
+        // First, we materialize the current metadata image and send it to all the publishers that
```
It would be helpful to add some motivation for this. Is it for recovery if something goes wrong after upgrade/downgrade?

cmccabe replied:
The issue is that different publishers may be installed at different times.

For example, let's say you have some hypothetical cluster balancing service that can be dynamically enabled. (not included in this PR, of course). When it's dynamically enabled, it registers as a metadata publisher, to get all the updates.

If you enable this thing well after startup, you will need to send it a MetadataDelta that contains all the updates (since all the updates are new to it, although not to the existing publishers). How do you get such a MetadataDelta? Write out the existing image to a delta.

Basically this PR adds dynamic installation / removal to the system, which it didn't previously have. That's why the old code locks it to a single publisher. But that isn't very dynamic and we have some use cases for installing things at runtime.

hachikuji commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -0,0 +1,504 @@
+            SnapshotReader<ApiMessageAndVersion> reader
+    ) {
+        long startNs = time.nanoseconds();
+        int snapshotIndex = 0;
```
Not sure how useful this index is given the presence of control records in the snapshot. Why not use the offset from `Batch` as we do when processing log records?

cmccabe replied:
Hmm. I'm actually a bit confused about what that field in the batch contains, since from the KRaft point of view all the records in a snapshot are at the same "offset"

Since snapshotIndex is ported from the old code, perhaps we can look into it a bit later? Or do a follow-on JIRA...

mumrah commented on metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java:
```diff
@@ -0,0 +1,262 @@
+
+    public void disable(String disabledReason) {
+        log.error("Disabling periodic KRaft snapshot generation because {}.", disabledReason);
+        this.disabledReason = disabledReason;
```
I think if we pass `null` here, it will re-enable snapshot generation. Do we want to allow that? Or should snapshot disabling be irreversible until restart?

cmccabe replied:
Yeah, I think we should not allow re-enablement. I will add a `requireNonNull`

mumrah commented on core/src/main/scala/kafka/server/KafkaRaftServer.scala:
```diff
@@ -81,26 +83,69 @@ class KafkaRaftServer(
+    build()
+
+  private val snapshotGenerator: SnapshotGenerator = new SnapshotGenerator.Builder().
+    setTime(Time.SYSTEM).
```
Should be the `time: Time` passed into KafkaRaftServer

mumrah commented on core/src/main/scala/kafka/server/KafkaRaftServer.scala:
```diff
@@ -81,26 +83,69 @@ class KafkaRaftServer(
+
+  private val metadataLoader: MetadataLoader = {
+    val builder = new MetadataLoader.Builder().
+      setTime(Time.SYSTEM).
```
Same comment as above ^

cmccabe committed:
some fixes

cmccabe committed:
address review comments

cmccabe committed:
Get snapshots tests with KafkaClusterTestKit working

- Rename SnapshotGenerator#minTimeSinceLastSnapshotNs to maxTimeSinceLastSnapshotNs. It is a
maximum, not a minimum.

- KafkaClusterTestKit should create a SnapshotGenerator if either the time-based or bytes-based
config is set (but not otherwise).

cmccabe committed:
DynamicConfigPublisher: ignore handlers that aren't present

Ignore handlers that aren't present (like the TOPIC handler on the
controller)

cmccabe committed:
Add testMetadataVersionChange

cmccabe committed:
KafkaClusterTestKit: fix thread name prefixes

cmccabe committed:
KRaftClusterTest: set higher values for max.connections.per.ip

cmccabe committed:
JointServer refactor

cmccabe committed:
SnapshotGenerator.java: add more log4j

cmccabe committed:
Some simplifications and be sure to register SnapshotGenerator in JointServer

cmccabe committed:
KafkaClusterTestKit: fix colocated mode

cmccabe committed:
Add EventQueue#size and make EventQueue#isEmpty take into account currently running events

cmccabe committed:
Re-fix KafkaClusterTestKit in combined mode

cmccabe committed:
Fix PlaintextAdminIntegrationTest.testCreateTopicsReturnsConfigs

cmccabe committed:
Merge branch 'trunk' into ak-image4

hachikuji commented on core/src/main/scala/kafka/server/KafkaRaftServer.scala:
```diff
@@ -69,95 +64,54 @@ class KafkaRaftServer(
   override def shutdown(): Unit = {
+    // The last component to be shut down will stop JointServer.
     broker.foreach(_.shutdown())
-    // The order of shutdown for `RaftManager` and `ControllerServer` is backwards
```
It looks like we lost this in the updated implementation? The socket server of the controller is shutdown before `JointServer`. By the time we get to `RaftManager` shutdown, the socket server is down and we cannot do graceful resignation from the quorum anymore.

hachikuji commented on core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:
```diff
@@ -0,0 +1,126 @@
+import org.apache.kafka.server.fault.FaultHandler
+
+
+class DynamicConfigPublisher(
```
Can we pull this into a separate PR? It is asking a lot from reviewers to be rigorous with a 4k diff. As I see it, there are at least 3 separate PRs here:

1. General refactor of server startup/shutdown
2. Implementation of time-based snapshots
3. Implementation of dynamic controller configuration

cmccabe replied:
Here's a PR for just JointServer and nothing else : https://github.com/apache/kafka/pull/12837

I would do split 2 and  split 3 differently, as "new code" versus "changes to old code." Let me know if that makes sense.

cmccabe commented:
The controller-side changes are now here: #12983

cmccabe commented:
I split this PR into a lot of PRs and they are now all in. So time to close this!

Thanks, all.
